### PR TITLE
Exposed engine must include all operations below global checkpoint during rollback

### PR DESCRIPTION
**Pull request by dnhatn**
Today we expose a new engine immediately during Lucene rollback. The new engine is started with a safe commit which might not include all acknowledged operation. With this change, we won't expose the new engine until it has recovered from the local translog.

Note that this solution is not complete since it's able to reserve only acknowledged operations before the global checkpoint. This is because we replay translog up to the global checkpoint during rollback. A per-doc Lucene rollback would solve this issue entirely.

Relates #32867

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left 2 smaller comments around the order of things and mutexes. Looking good otherwise

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -1378,19 +1385,30 @@ private void innerOpenEngineAndTranslog() throws IOException {
...
-        createNewEngine(config);
-        verifyNotClosed();
-        // We set active because we are now writing operations to the engine; this way, if we go idle after some time and become inactive,
-        // we still give sync'd flush a chance to run:
-        active.set(true);
+        Engine newEngine = null;
+        try {
+            assert currentEngineReference.get() == null : "engine is running";
+            trimUnsafeCommits();
+            newEngine = engineFactory.newReadWriteEngine(config);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: We [discussed](https://github.com/elastic/elasticsearch/pull/32867#discussion_r212170675) this before but somehow I missed it. We must create a new engine under the mutex. I have updated this.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -2793,21 +2786,44 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
...
+
+        Engine newEngine = null;
+        try {
+            final long globalCheckpoint = getGlobalCheckpoint();
             trimUnsafeCommits();
-            newEngine = createNewEngine(newEngineConfig());
-            active.set(true);
+            newEngine = engineFactory.newReadWriteEngine(newEngineConfig());
+            onNewEngine(newEngine);
+            newEngine.onSettingsChanged();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Good catch. I have updated this.

**Comment by dnhatn** [comment]
@ywelsch Two very good catches. It's ready for another round. Would you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -2852,21 +2842,46 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
     void resetEngineToGlobalCheckpoint() throws IOException {
         assert getActiveOperationsCount() == 0 : "Ongoing writes [" + getActiveOperations() + "]";
         sync(); // persist the global checkpoint to disk
-        final long globalCheckpoint = getGlobalCheckpoint();
-        final Engine newEngine;
+        final SeqNoStats seqNoStats = seqNoStats();
+        final TranslogStats translogStats = translogStats();
+        // flush to make sure the latest commit, which will be opened by the read-only engine, includes all operations.
+        flush(new FlushRequest());
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: do we need to set `waitIfOngoing` to be 100% sure it's finished?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -2852,21 +2842,46 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
...
+        Engine newEngine = null;
+        try {
+            final long globalCheckpoint = getGlobalCheckpoint();
             trimUnsafeCommits();
-            newEngine = createNewEngine(newEngineConfig());
-            active.set(true);
+            synchronized (mutex) {
+                // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
+                newEngine = engineFactory.newReadWriteEngine(newEngineConfig());
+                onNewEngine(newEngine);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: onNewEngine does't publish the new engine right?

**Review by s1monw — APPROVED** [review]
left 2 comments. LGTM otherwise

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 5]
```diff
@@ -2852,21 +2842,46 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
...
+        }
+
+        Engine newEngine = null;
+        try {
+            final long globalCheckpoint = getGlobalCheckpoint();
             trimUnsafeCommits();
-            newEngine = createNewEngine(newEngineConfig());
-            active.set(true);
+            synchronized (mutex) {
+                // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: ++

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Yes, I pushed https://github.com/elastic/elasticsearch/pull/36159/commits/a907a8e4214281528575379a6d4b3c51256a0b39

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Yes, I pushed https://github.com/elastic/elasticsearch/pull/36159/commits/a907a8e4214281528575379a6d4b3c51256a0b39.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: `onNewEngine` does not expose the engine itself but exposes only the last refresh translog location to RefreshListeners.

**Pull request merged**